### ISSUE TITLE
Allow the execution of `initStatefulProperties` for non-stateful materials

### ISSUE THREAD

**Issue by hugary1995** [idaholab/moose#28971]
### Motivation

Some of my old input files no longer work. I've done some digging and found that the root cause is the following pattern
```python
[Materials]
  [props]
    type = GenericConstantMaterial
    prop_names = 'a0'
    prop_values = '1'
  []
  [stateful_props]
    type = MyMaterial
    initial_condition = 'a0'
  []
[]
```
I gather that now if none of the declared material properties are stateful, the material object's initStatefulProperties method is not called at all (need someone to confirm), which is great. But how about the above use case where I want to use the non-stateful material property to initialize a stateful property? An apparent workaround is to promote "a0" to be stateful in MyMaterial, but that will lead to additional memory consumption.

### Design

Quoting @permcody 
> I feel like we should be able to support this without the workaround, which could be expensive.

So let's think about it... The end goal is simply:
1. Smartly figure out when we should call `initStatefulProperties` even if a material object does not provide any stateful properties
2. While we do 1, make sure we don't _actually_ make those properties stateful and hence consume more memory

2 is relatively easy to address -- we just treat those properties as non-stateful properties in the material storage.

1 could be tricky, and the key is how to "smartly" detect the case. A possible approach is to add an option to the `getMaterialProperty` APIs to mark the retrieved property as "not stateful but is depended by one or more stateful properties".

### Impact

The proposed solution will introduce an additional optional argument in the `getMaterialProperty` APIs. Addressing this issue will enable the custom material design pattern mentioned above.

**Comment by YaqiWang** [comment]
I am curious that we do have evaluations on initial, in which the material containing stateful properties can use the non-stateful material properties, is that a solution for you?

**Comment by hugary1995** [comment]
Is that really happening, though? If that is the expected behavior, there is a bug. I can prepare a MRE if you confirm.

In other words, I don't think the value of a0 is computed when I'm initializing my stateful properties in `initQpStatefulProperties`.

Or are you suggesting something like
```cpp
void
MyMaterial::computeQpProperties()
{
  if (_step == 0)
    _a[_qp] = _a0[_qp];
  else
  {
    // ...
  }
}
```
This will work, but I would still call it a workaround, not a solution to this problem.

**Comment by hugary1995** [comment]
I think if we educate users and new developers to use `initQpStatefulProperties` to initialize stateful properties, we should try to stick with that paradigm and handle edge cases (by us, not by users).

**Comment by YaqiWang** [comment]
Yes, I was suggesting that `MyMaterial::computeQpProperties()` with if statement. I would not call it as a workaround though. I refer to it as initial evaluation vs initialization in `initQpStatefulProperties`. Aux variables have the same situation. The general setup may contain coupled variable for evaluating a non-stateful material property, which can only be resolved with full machinery in `FEProblemBase::execute`.

**Comment by hugary1995** [comment]
I think I understand your arguments, but allow me to disagree, and the reason is simple: If there are so many cases that `initStatefulProperties` cannot handle (while the `computeQpProperties` and if-statement approach can), we should entirely get rid of the `initStatefulProperties` API.

**Comment by YaqiWang** [comment]
;-) if you can agree with deleting `initStatefulProperties`, `computeQpProperties` with if statement is acceptable to you. In my opinion, `initStatefulProperties` should only do something like what initial conditions do to aux variables. Having this function can possibly make stateful properties available earlier. In the code I am working on, we have a material having both but the two cannot function at the same time. This issue could be also resolved by merging two materials together.

**Comment by hugary1995** [comment]
> In my opinion, `initStatefulProperties` should only do something like what initial conditions do to aux variables like setting constant values or values based on a function, etc.

The issue is that we can't prevent users from using non-stateful properties to initialize a stateful property, can we? This leads to the mistake I made, which silently yields wrong results -- the worst possible outcome.

**Comment by YaqiWang** [comment]
>silently yields wrong results

Can valgrind detect this?

**Comment by hugary1995** [comment]
LOL, possibly, if non-stateful properties are truly non-initialized at that stage...

**Comment by hugary1995** [comment]
Chances are they are initialized to 0, IIRC. So unless we change that behavior, valgrind may not be able to detect it.

**Comment by YaqiWang** [comment]
Right. Whether initialize the memory after declaration could be debatable. I think we should not initialize. This is off the topic a little.
